### PR TITLE
Use dedicated sccache GHA support + Linux support

### PR DESCRIPTION
**Pull request by quyykk**
This PR uses sccache's dedicated GHA cache support, and changes the Linux build to use sccache as well.

From what I can tell (testing it https://github.com/quyykk/endless-sky/pull/6), it is even faster than our current caching workflow (50% approx). I don't know why this is the case, but it is.

It also changes our caching strategy: Currently, PR changes are cached as well. This is pretty wasteful, and in fact we regularly hit our cache quota of 10GB (https://github.com/endless-sky/endless-sky/actions/caches). So, this PR only saves caches by actual commits to the repo, and not PRs. This does mean that PRs that change a lot of files won't be able to greatly benefit from the cache, but oh well.

I have implemented MacOS support too, but it is currently unused because the xcode toolchain cannot make use of sccache.

Another improvement compared to our current caching workflow is that the CD runs are cached as well!

### CODE REVIEW

File: .github/sccache/action.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    - name: Download sccache (Windows)
+      if: runner.os == 'Windows'
+      run: gh release download -R mozilla/sccache ${{ env.SCCACHE_VERSION }} -p sccache-${{ env.SCCACHE_VERSION }}-x86_64-pc-windows-msvc.tar.gz
+      shell: bash
+
+    - name: Download sccache (MacOS)
+      if: runner.os == 'macOS'
+      run: gh release download -R mozilla/sccache ${{ env.SCCACHE_VERSION }} -p sccache-${{ env.SCCACHE_VERSION }}-x86_64-apple-darwin.tar.gz
+      shell: bash
```
Comment (MCOfficer) [inline .github/sccache/action.yml:30]: What do you think about moving the if-else into bash, to reduce the amount of duplicated code?
```suggestion
    - name: Download sccache
      if: runner.os == 'Linux'
        run: |
          if [ $RUNNER_OS == "Linux"]; then
              PLATFORM="unknown-linux-musl"
          elif [ $RUNNER_OS == "Windows"]; then
              PLATFORM="pc-windows-msvc"
          elif [ $RUNNER_OS == "macOS"]; then
              PLATFORM="apple-darwin"
          fi
          gh release download -R mozilla/sccache ${{ env.SCCACHE_VERSION }} -p sccache-${{ env.SCCACHE_VERSION }}-x86_64-$PLATFORM.tar.gz
      shell: bash
```

File: .github/sccache/action.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    - name: Extract
+      run: tar xf sccache-*.tar.gz --strip-components=1
+      shell: bash
+    - name: Add to PATH
+      run: echo "." >> $GITHUB_PATH
```
Comment (MCOfficer) [inline .github/sccache/action.yml:36]: Paths added to PATH should ideally be absolute, f.e. `$PWD`

File: .github/workflows/cd.yaml [hunk 3]
```diff
@@ -30,6 +31,11 @@ jobs:
           sudo apt-get update
           sudo apt-get install -y --no-install-recommends libsdl2-dev libpng-dev libjpeg-turbo8-dev libopenal-dev libmad0-dev libglew-dev libgl1-mesa-dev libegl1-mesa-dev libgles2-mesa-dev uuid-dev
           python3 -m pip install --user scons
+      - name: Setup sccache
+        uses: ./.github/sccache
+        with:
+          read-only: 'true'
+          cache-key: ubuntu-20.04-desktop
```
Comment (MCOfficer) [inline .github/workflows/cd.yaml:38]: Can you unify the cache keys to some sensible, shared format? right now it's a mix of hardcoded strings, `{{ matrix.os }}` and `${{ matrix.os }}-desktop`

**Review by MCOfficer — CHANGES_REQUESTED** [review]

File: .github/workflows/cd.yaml (hunk 3 above)
Comment (MCOfficer) [reply .github/workflows/cd.yaml:38]: what about `{{ runner.os }}-{{ runner.arch }}`?

As for `-desktop`, i'm fine with leaving it there but perhaps make it clearer that this is about GL being present. perhaps `-headless` for those without.

Comment (quyykk) [reply .github/workflows/cd.yaml:38]: `runner.os` doesn't include the OS version :confused: So I can't use that unfortunately.

I could rename `desktop` to `opengl`? would that help? I don't think `headless` makes sense (just like `desktop` makes no sense either).

Comment (MCOfficer) [reply .github/workflows/cd.yaml:38]: > `runner.os` doesn't include the OS version 😕 So I can't use that unfortunately.

why is that relevant?

> I could rename `desktop` to `opengl`? would that help? I don't think `headless` makes sense (just like `desktop` makes no sense either).

good enough

Comment (quyykk) [reply .github/workflows/cd.yaml:38]: > why is that relevant?

I guess it's only relevant for the Ubuntu jobs (20 vs 22).

Comment (MCOfficer) [reply .github/workflows/cd.yaml:38]: But even then - what is the difference? The only big kicker i can think of is dependency versions, and us caching the compilation of *our* code shouldn't be relevant to that.

I mean, at that point it's worth asking why we have jobs for both ubuntu versions in the first place. Unless it's explicitly to test everything on both versions, we might as well share the cache between them.

Comment (quyykk) [reply .github/workflows/cd.yaml:38]: Compiler version too. From my perspective I'm fine with just having one Ubuntu version

**Review by MCOfficer — APPROVED** [review]

**Pull request merged**